### PR TITLE
Retry in host.json should apply to C# pre-compiled functions (#7232)

### PR DESCRIPTION
**Pull request by alrod**
Retry in host.json should apply to C# pre-compiled functions (#7232)

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [x] Otherwise: Documentation issue linked to PR
* [] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

Corresponding web jobs PR:
Azure/azure-functions-host/issues/7232

### CODE REVIEW

**Review by pragnagopa — DISMISSED** [review]

**Comment by pragnagopa** [comment]
@mathewc  - LGTM - Please take a look as well to ensure as this a breaking change w.r.t abstractions package

File: src/WebJobs.Script/Config/RetryOptionsSetup.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        public RetryOptionsSetup(IConfiguration configuration)
+        {
+            _configuration = configuration;
+        }
+
+        public void Configure(RetryOptions options)
+        {
+            IConfigurationSection jobHostSection = _configuration.GetSection(ConfigurationSectionNames.JobHost);
+            var healthMonitorSection = jobHostSection.GetSection(ConfigurationSectionNames.Retry);
```
Comment (mathewc) [inline src/WebJobs.Script/Config/RetryOptionsSetup.cs]: ? health monitor?

File: src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs [hunk 2]
```diff
@@ -42,11 +42,6 @@ public FunctionMetadata()
         /// </summary>
         public IDictionary<string, object> Properties { get; }
 
-        /// <summary>
-        /// Gets or sets the optional function execution retry strategy to use on invocation failures.<see cref="Abstractions.Description.RetryOptions"/>.
-        /// </summary>
-        public RetryOptions Retry { get; set; }
```
Comment (mathewc) [inline src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs:48]: I thought we had told logic apps to use this as a way to specify retry for their dynamically generated functions? I.e. this breaking change might affect them. They might not have taken a dependency on this yet though - we just need to check.

If they haven't, I buy your argument that it's probably safe for us to make this change, given they're really our only consumer right now of this preview package.

**Review by liliankasem — DISMISSED** [review]

File: src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <Import Project="..\..\build\common.props" />
     <PropertyGroup>
-    <Version>1.0.3-preview</Version>
```
Comment (mathewc) [inline src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj:4]: We've never even released v1 and it's still in preview. Seems strange to be bumping the major version of it if it's never been released. @fabiocav for visibility

Comment (pragnagopa) [reply src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj:4]: I recommended bumping major version as  Logic Apps team has dependency on this package.

File: src/WebJobs.Script/Config/RetryOptionsSetup.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Script.Configuration;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.Config
+{
+    internal class RetryOptionsSetup : IConfigureOptions<RetryOptions>
```
Comment (mathewc) [inline src/WebJobs.Script/Config/RetryOptionsSetup.cs:11]: I didn't see a test for this options setup verifying that it correctly translates json into a RetryOptions

File: src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs (hunk 2 above)
Comment (alrod) [reply src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs:48]: Chatted with LogicApps offline - they do not use Retry. So we should be good.

File: src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj (hunk 3 above)
Comment (mathewc) [reply src/WebJobs.Script.Abstractions/WebJobs.Script.Abstractions.csproj:4]: I think we should just increment this normally to 1.0.4-preview. Bumping the major version wouldn't prevent issues if extensions are actually using this I think. E.g. say an extension is using the existing 1.0.3 version and setting a retry. An updated host using version 2.0.0 wouldn't see/apply that retry anymore (breaking). Depending on how assembly unification works in this case, the issue might even be worse - if the extension unifies to the host version, the extension would see missing method exceptions?

File: src/WebJobs.Script/Extensions/FunctionMetadataExtensions.cs [hunk 5]
```diff
@@ -89,11 +91,29 @@ internal static void SetFunctionId(this FunctionMetadata metadata, string functi
...
+            GetObjectProperty(metadata.Properties, RetryKey) as RetryOptions;
+
         private static bool GetBoolProperty(IDictionary<string, object> properties, string propertyKey)
         {
             return properties.TryGetValue(propertyKey, out object valObj)
                 && valObj is bool valBool
                 && valBool;
         }
+
+        private static object GetObjectProperty(IDictionary<string, object> properties, string propertyKey)
```
Comment (mathewc) [inline src/WebJobs.Script/Extensions/FunctionMetadataExtensions.cs]: What value is this method providing? The underlying dictionary already returns objects, so I'd just code directly to the dictionary in GetRetry above

**Review by mathewc — DISMISSED** [review]

**Comment by liliankasem** [comment]
Closing this PR as suggested here https://github.com/Azure/azure-webjobs-sdk/pull/2786

"As we are not going to have host level support for retries the commit can be reverted and the host PR can be closed"

**Pull request closed without merging**